### DISCUSSION TITLE
How to use Ogma logger with Nest Commander in `main.ts`?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [jmcdo29/nest-commander#884] [Q&A] [answered]
In `CommandFactory.run()`, I can add `CommandFactoryRunOptions` object, however, `logger` property accepts only `Logger` from `@nestjs/logger`.

**Comment by jmcdo29** [comment, marked as answer]
Any logger that satisfies Nest's `logger` property is valid, however using ogma directly for that isn't always nice to have to instantiate the `Ogma` instance yourself, so instead with v3.7.0 you can do something like
```ts
async function bootstrap() {
  const app = await CommandFactory.createWithoutRunning(AppModule);
  app.useLogger(app.get(OgmaService)); // assuming `OgmaModule.forRoot()` has been imported
  await CommandFactory.runApplication(app);
}
bootstrap();
```
[Docs on the createWithoutRunning](https://nest-commander.jaymcdoniel.dev/en/features/factory/#creating-an-application-without-running-it)

**Reply by tukusejssirs** [reply]
In the app module, what should be the `interceptor` value? Or should I simply omit it? :thinking:

**Reply by jmcdo29** [reply]
You can omit it or set it to `false`. Either should work. `nest-commander` is a [standalone application](https://docs.nestjs.com/standalone-applications) which means that enhancers do not get bound.

**Reply by tukusejssirs** [reply]
Okay, I tried this out. My app works when I add this to `main.ts` and when I define `OgmaModule.forRoot()` and `OgmaModule.forFeatures()` in the module file.

However, when I add `constructor(@OgmaLogger(SomeCommand.name) private readonly logger: OgmaService) {}` into the command file, TS errors out with `TS2377: Constructors for derived classes must contain a 'super' call.`, so I add `super()` into the constructor (i.e. `constructor(...) {super()}`), but then the app does not work: it can be built and run, but the exit code is always `1`. I have no idea how to get debug this (even when I use `NODE_DEBUG='*'`, no error is thrown). :thinking:

**Reply by jmcdo29** [reply]
Got a reproduction? For sake of getting debug logs you can always do `logger: new ConsoleLogger()` to get Nest's default output

**Reply by tukusejssirs** [reply]
[Here](https://github.com/ts-tests/comm-ogma)’s the reproduction.

Steps:

1. Clone.
2. Install NPM dependencies.
3. Build (`pnpm build`).
4. `pnpm should:fail` fails with exit code `1` (no log from the app itself). This is not expected.
5. `pnpm should:succeed` fails with exit code `1` (no log from the app itself). This is not expected.
6. In [`src/cli/command/plc.command.ts`](https://github.com/ts-tests/comm-ogma/blob/main/src/cli/command/plc.command.ts), comment out lines 18 and 32.
7. Re-build the app.
8. `pnpm should:fail` outputs `'not_an_ip_address' is not a valid IP address.` and exits with `2`. This is expected.
9. `pnpm should:succeed` outputs `{ inputs: [], options: { ip: '1.1.1.1' } }` and exits with `0`. This is expected.

Now, I have no idea why the output and exit codes of steps 4 and 5 and not the same as of steps 8 and 9 respectively (of course, the log on L32 should be the only difference).

**Reply by jmcdo29** [reply]
Move `PlcCommand` to the `providers` array

**Reply by tukusejssirs** [reply]
Oh, such a such newbie mistake! :man_facepalming: Thanks! :pray:
